zerver/views/streams.py

```diff
@@ -525,20 +539,88 @@ def you_were_just_subscribed_message(
     if len(subscriptions) == 1:
         with override_language(recipient_user.default_language):
             return _("{user_full_name} subscribed you to the stream {stream_name}.").format(
-                user_full_name=f"@**{acting_user.full_name}|{acting_user.id}**",
+                user_full_name=silent_mention_syntax_for_user(acting_user),
                 stream_name=f"#**{subscriptions[0]}**",
             )
 
     with override_language(recipient_user.default_language):
         message = _("{user_full_name} subscribed you to the following streams:").format(
-            user_full_name=f"@**{acting_user.full_name}|{acting_user.id}**",
+            user_full_name=silent_mention_syntax_for_user(acting_user),
         )
     message += "\n\n"
     for stream_name in subscriptions:
         message += f"* #**{stream_name}**\n"
     return message
 
 
+def you_were_just_unsubscribed_message(
+    acting_user: UserProfile,
+    recipient_user: UserProfile,
+    stream_names: Set[str],
+) -> str:
+    subscriptions = sorted(stream_names)
+    if len(subscriptions) == 1:
+        with override_language(recipient_user.default_language):
+            return _("{user_full_name} unsubscribed you from the stream {stream_name}.").format(
+                user_full_name=silent_mention_syntax_for_user(acting_user),
+                stream_name=f"#**{subscriptions[0]}**",
+            )
+
+    with override_language(recipient_user.default_language):
+        message = _("{user_full_name} unsubscribed you from the following streams:").format(
+            user_full_name=silent_mention_syntax_for_user(acting_user),
+        )
+    message += "\n\n"
+    for stream_name in subscriptions:
+        message += f"* #**{stream_name}**\n"
+    return message
+
+
+def send_messages_for_removed_subscribers(
+    removed_subscribers: List[Tuple[UserProfile, Stream]], acting_user: UserProfile
+) -> None:
+    if len(removed_subscribers) == 0:
+        return
+
+    user_id_to_unsubscribed_streams: Dict[int, Set[str]] = {}
+    user_id_to_user: Dict[int, UserProfile] = {}
+    for user, stream in removed_subscribers:
+        if user.id in user_id_to_unsubscribed_streams:
+            user_id_to_unsubscribed_streams[user.id].add(stream.name)
+        else:
+            user_id_to_unsubscribed_streams[user.id] = {stream.name}
+            user_id_to_user[user.id] = user
+
+    mention_backend = MentionBackend(acting_user.realm_id)
+    message_requests = []
+    for user_id, stream_names in user_id_to_unsubscribed_streams.items():
+        # Our caller already filters deactivated users. We need to check for
+        # other cases where we shouldn't send notification message.
+        if user_id == acting_user.id:
+            continue
+        if user_id_to_user[user_id].is_bot:
+            continue
+
+        sender = get_system_bot(settings.NOTIFICATION_BOT, user_id_to_user[user_id].realm_id)
```

Comment:
We can avoid fetching sender in loop to reduce db query. The sender would be the same here, right?

Reply:
Yup! Our caller ensures that all users are in the same realm. Made the following change, though this did not reduce query count, it did reduce the cache count (as expected) in `test_subs.py`.

```diff
--- a/zerver/views/streams.py
+++ b/zerver/views/streams.py
@@ -582,6 +582,7 @@ def send_messages_for_removed_subscribers(
             user_id_to_user[user.id] = user

     mention_backend = MentionBackend(acting_user.realm_id)
+    sender: Optional[UserProfile] = None
     message_requests = []
     for user_id, stream_names in user_id_to_unsubscribed_streams.items():
         # Our caller already filters deactivated users. We need to check for
@@ -591,7 +592,12 @@ def send_messages_for_removed_subscribers(
         if user_id_to_user[user_id].is_bot:
             continue

-        sender = get_system_bot(settings.NOTIFICATION_BOT, user_id_to_user[user_id].realm_id)
+        # Our caller ensures that acting_user and all users that they unsubscribe are in
+        # the same realm. This allows us to only fetch this bot once and avoid having to
+        # fetch again with unsubscribed user's realm_id on every iteration of the loop.
+        if sender is None:
+            sender = get_system_bot(settings.NOTIFICATION_BOT, acting_user.realm_id)
+
         message = you_were_just_unsubscribed_message(
             acting_user=acting_user,
             recipient_user=user_id_to_user[user_id],
```